### ISSUE TITLE
[API Proposal] AllowListSerializationBinder

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/runtime#104037]
## Background and Motivation

In .NET 9 we have moved `BinaryFormatter` to an OOB package. Some users are still going to use it and they are going to remain vulnerable.

To improve the security, the users should implement a custom `SerializationBinder` and limit the allowed types to a known, closed set of types.

Implementing such binder seems very simple at the beginning, but in the reality it's hard to do right ([example](https://github.com/dotnet/runtime/blob/5c53e8ec9ca1a06a31c45936d3058fd6b1dc5961/src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameParserSamples.cs#L75-L81)).

## Proposed API

```csharp
namespace System.Runtime.Serialization;

public sealed class AllowListSerializationBinder : SerializationBinder
{
    public AllowListSerializationBinder(params Type[] allowedTypes);
    
    public override Type BindToType(string assemblyName, string typeName);
}
```

`allowedTypes` would not need to contain any of the most popular primitive types listed [here](https://github.com/dotnet/runtime/blob/5c53e8ec9ca1a06a31c45936d3058fd6b1dc5961/src/libraries/System.Reflection.Metadata/tests/Metadata/TypeNameParserSamples.cs#L24-L42).

## Usage Examples

```csharp
AllowListSerializationBinder s_binder = new(allowedTypes:
[
    typeof(CustomUserDefinedType),
    typeof(List<>) // allow for List<T> where T is CustomUserDefinedType or any of the always-enabled primitive types
]);

static Deserialize<T>(Stream bfStream)
{
    BinaryFormatter bf = new()
    {
        Binder = s_binder
    };

    return (T)bf.Deserialize(bfStream);
}
```

## Risks

Providing such feature may give our users a false impression that using `BinaryFormatter` can be safe.

cc @GrabYourPitchforks @bartonjs

**Comment by adamsitnik** [comment]
cc @jkotas

**Comment by stephentoub** [comment]
> Providing such feature may give our users a false impression that using BinaryFormatter can be safe.

Is it "safe" if one uses this binder? If yes, why did we just go through the enormous effort (which will have massive ripples for a long time) of extricating BinaryFormatter? And if no, why would we add such a new type to a vulnerable package when it doesn't actually make it safe? It makes it "enough safe" that it's worth it?

**Comment by jkotas** [comment]
I agree with @stephentoub . We should be doing everything we can do discourage binary serializers.

> in the reality it's hard to do righ

I see this as a good thing. It discourages people from using binary serializers.

**Comment by adamsitnik** [comment]
> Is it "safe" if one uses this binder?

It solves only a subset of known issues, but will never be 100% safe.

Since all architects are against, I am just going to close it. I am also going to remove the part that mentions it in the migration doc.
